### ISSUE TITLE
[EPIC] Allow Downloading all attachment as once

### ISSUE THREAD

**Issue by chibenwa** [linagora/tmail-backend#629]
As a user I want to have a button in TeamMail to downlaod all attachments of an email at once.

https://user-images.githubusercontent.com/6928740/229980634-4e25ae71-814a-4b39-ae93-e08484c7cf8f.mp4

We will design and implement a JMAP extension (download all):

```
POST download

[
    {
        "blobId": "abcd",
        "filename": "star.png"
    },
    {
        "blobId": "defg",
        "filename": "planet.png"
    },
    ...
]
```

Will return a ZIP containing all the files.

 -> Reject if `not found`, `no right to access`, `too big`
 
 - [ ] ADR
 - [ ] Document the JMAP extension
 - [ ] Groom & implement

**Comment by Arsnael** [comment]
Implementing and using `Blob/get` for example from https://www.ietf.org/archive/id/draft-ietf-jmap-blob-18.html#name-blob-get that is in the rfc editor queue and bound to be soon released as a standard wouldn't get the same result, without having to develop a custom TMail method for that?

**Comment by chibenwa** [comment]
Blob/get is base64 encoded on a JSON

Not appropriate for big downloads ;-)

**Comment by chibenwa** [comment]
This point is required by MU so let's start working on this...

Grooming first, please!

**Comment by chibenwa** [comment]
We need a JMAP capability too...

**Comment by Arsnael** [comment]
> This point is required by MU so let's start working on this...

ETA on this? as Grooming takes a bit of time and tet is around the corner?

**Comment by Arsnael** [comment]
https://github.com/linagora/james-project-private/issues/961

**Comment by chibenwa** [comment]
ETA: ~1 month

Anyway it is not that of a complicated topic.

I propose to implement a new JMAP capability: `com:linagora:params:downloadAll`

This capability will expose a new endpoint allowing to download all attachments of an EmailBodyPart (as referenced by `Email/get` and `Email/parse` (for nested emails)) in a ZIP with the actual files.

Edge cases: two files have the same filename: keep the latest one?

```
GET /jmap/downloadAll/{accountId}/{blobId}?name={name}

[ZIP DATA]
```

**Comment by quantranhong1999** [comment]
> Edge cases: two files have the same filename: keep the latest one?

It seems Gmail adds postfix `_1`, `_2`... after the filename and before the extension part.

E.g. Download two files with the same `check.sh` name return:
- `check.sh`
- `check_1.sh`

https://github.com/user-attachments/assets/20f4b992-e57d-4dce-b2cf-86b5e1be4ad3

**Comment by quantranhong1999** [comment]
> GET /jmap/downloadAll/{accountId}/{blobId}?name={name}

Should `blobId` be `emailId` instead?

**Comment by Arsnael** [comment]
Mainly agree with @quantranhong1999 here.

If multiple files have the same name we can add a suffix to it to differentiate.

Also blobId should be emailId. we downloads all attachments (blobIds) of an email.

List:
- ADR
- JMAP custom method documentation
- JMAP downloadAll route implementation with jmap capability `com:linagora:params:downloadAll`

That should be enough?

**Comment by chibenwa** [comment]
> It seems Gmail adds postfix _1, _2... after the filename and before the extension part.

Careful, let's handle this in a separate issue please if relevant. That's a bahaviour change.

> (Updated: likely implicitly the same as blobId can be emailId).

Indeed

**Comment by vttranlina** [comment]
How about idea to pre-upload single zip? (do it when save email)

Benefits:
- Reduced server load when download, (highly beneficial when emails with multiple recipients)
- Faster download: (in case we return signed url from S3, then we don't need forward traffic to Tmail in middle)
- Simpler client interaction (jmap client just send only emailId)
- We can use method `Blob/get` for it, with "data" property is a raw download url. (Not sure it is possible with current code)

Trade-offs
- Cost for S3 storage
- More Cpu load during upload
- Users cannot download "selected" attachments. (It is not a real user case, similar Gmail case)

**Comment by chibenwa** [comment]
> How about idea to pre-upload single zip? (do it when save email)

No no no no no

Don't impact 100% of the user (by doubling disk space!) for a side used 1% of the time.

**Comment by vttranlina** [comment]
> Don't impact 100% of the user (by doubling disk space!) for a side used 1% of the time.

We only upload case have >=2 attachment

**Comment by Arsnael** [comment]
I agree with @chibenwa here, I do not think it's a good idea. That could significantly increase the cost of S3 storage.

I don't think this is used too often as well as a function, so the performance cost should be minimal and more than acceptable

**Comment by vttranlina** [comment]
Note
Library for zip

```
            <dependency>
                <groupId>org.apache.commons</groupId>
                <artifactId>commons-compress</artifactId>
                <version>1.26.2</version>
            </dependency>
```
Code sample: 
`<artifactId>apache-james-mailbox-deleted-messages-vault</artifactId>`
and `<artifactId>backup</artifactId>`

**Comment by chibenwa** [comment]
## Definition of done

Performance: Given a message with N attachments, the message is actually downloaded once.

**Comment by Arsnael** [comment]
Written tickets (feel free to comment or add others if necessary):

- https://github.com/linagora/tmail-backend/issues/1467
- https://github.com/linagora/tmail-backend/issues/1468
- https://github.com/linagora/tmail-backend/issues/1469
- https://github.com/linagora/tmail-backend/issues/1471
- https://github.com/linagora/tmail-backend/issues/1472

**Comment by chibenwa** [comment]
Looks overall great.

Feel free to have a look to `mailbox-backup` in order to see how the ZIP can be used.
